### PR TITLE
Set rrtmgp default backend back to kokkos

### PR DESCRIPTION
**Pull request by jgfouca**
Non-determinism and poor performance have been resolved.

### CODE REVIEW

**Comment by ndkeen** [comment]
I'm unable to build with this branch.

```
/mscratch/sd/n/ndk/repos/jgfouca_rrtmgp_interface_back_to_kokkos/components/eamxx/../eam/src/physics/rrtmgp/external/cpp/rrtmgp/mo_gas\
_optics_rrtmgp.h(1264): error: namespace "Kokkos::Experimental" has no member "OffsetView"
    using oview_t = Kokkos::Experimental::OffsetView<T, LayoutT, DeviceT>;
                                          ^

/mscratch/sd/n/ndk/repos/jgfouca_rrtmgp_interface_back_to_kokkos/components/eamxx/../eam/src/physics/rrtmgp/external/cpp/rrtmgp/mo_gas\
_optics_rrtmgp.h(1264): error: expected a ";"
    using oview_t = Kokkos::Experimental::OffsetView<T, LayoutT, DeviceT>;
```

**Review by bartgol — DISMISSED** [review]

**Comment by AaronDonahue** [comment]
@jgfouca , it looks like all the AT fails are unit tests w/ rrtmgp in them.  Are these expected fails?  Can we merge?

**Comment by jgfouca** [comment]
@AaronDonahue , not yet. I goofed pretty badly and did a lot of testing with Kokkos off (so yakl rrtmgp). I'm seeings loads of problems now that I corrected that.

**Comment by AaronDonahue** [comment]
okay cool, just checking as I do my audit of the open PRs

**Comment by jgfouca** [comment]
OK, I think this is finally ready.

**Review by bartgol — APPROVED** [review]

**Review by bartgol — DISMISSED** [review]

**Comment by AaronDonahue** [comment]
@jgfouca , can you confirm that the fails we are seeing on `weaver` in the AT are expected?  Then you can manually merge.

**Comment by jgfouca** [comment]
@AaronDonahue , that baseline_cmp fail on weaver is expected. @ndkeen and I are investigating potential issues with the new rrtmgp using more memory than before (with YAKL). I can merge now or hold off until we know more.

**Comment by bartgol** [comment]
@jgfouca I set WIP so the AT stops testing (it retests if master changes) until you decide on merging.

Regarding the memory: how much more memory are we talking about? If it's not too much more, we may just merge.

**Comment by AaronDonahue** [comment]
@jgfouca , i'm happy to leave the merge decision up to you.  I'm happy to wait while you investigate w/ Noel, or merge it.

**Comment by jgfouca** [comment]
@bartgol , I'm not sure exactly how much more memory, but it's enough the @ndkeen says ne120 no longer fits on 8 nodes on pm-gpu. He also reported some huge (3x) slowdowns in rrtmgp.

File: components/eamxx/src/physics/rrtmgp/scream_rrtmgp_interface.hpp
Comment (bartgol) [inline components/eamxx/src/physics/rrtmgp/scream_rrtmgp_interface.hpp:243]: What's base_ref used for?

Comment (jgfouca) [reply components/eamxx/src/physics/rrtmgp/scream_rrtmgp_interface.hpp:243]: This is a bit hacky. When I was doing the standalone rrtmgp work, I tried to come up with a simple function for the amount of pool memory needed. I found that comparing (ncol * nlay * nlev) to base_ref got me a reasonable estimate, but for some reason, way less is needed for full cases.

Comment (bartgol) [reply components/eamxx/src/physics/rrtmgp/scream_rrtmgp_interface.hpp:243]: this is just the "initial" guess for the pool, and then it can be expanded if needed, right?

Comment (jgfouca) [reply components/eamxx/src/physics/rrtmgp/scream_rrtmgp_interface.hpp:243]: Not sure what you mean. We can change this number in the source code (which reqs a rebuild to take effect). The pool cannot expand at runtime; so if we overflow it, the case will crash.

Comment (bartgol) [reply components/eamxx/src/physics/rrtmgp/scream_rrtmgp_interface.hpp:243]: Ah, ok. I thought it was like YAKL, which whenever the pool fills, opens another pool.

Is there any back of the envelope calculation to ensure this size is enough then?

Comment (jgfouca) [reply components/eamxx/src/physics/rrtmgp/scream_rrtmgp_interface.hpp:243]: @bartgol , I'm not sure YAKL works that way. There's been many times we've had to increase `GATOR_INITIAL_MB`.

I arrived at this number through experimentation on rrtmgp standalone.

**Review by bartgol — DISMISSED** [review]

**Comment by ndkeen** [comment]
A quick profile with ne30 on pm-gpu using new kokkos branch, I'm seeing a few CudaMalloc/CudaFrees that are most likely a great deal of the slowdown.
Screenshot 2024-11-07 at 1 24 41 PM

**Comment by jgfouca** [comment]
@ndkeen , thanks! Is there a way to get a line number or kernel name to see where these mallocs are happening?

**Comment by ndkeen** [comment]
No, the profiler uses sampling. Previously, these CudaMallocs are from Kokkos views.

**Comment by jgfouca** [comment]
@ndkeen , that is so weird. Can you confirm that you are on the correct rrtmgp sha? If you look at the rrtmgp file `mo_gas_optics_rrtmgp.h` and the kokkos impl of function `gas_optics` (around line 2000), near the top of the function you should see the pool allocator being used. I also hope you are on the most recent sha of this branch.

Within the main rrtmgp impl function, `RRTMGPRadiation::run_impl` (in `eamxx_rrtmgp_process_interface.cpp`), there should not be any cuda mallocs or frees happening.

**Comment by ndkeen** [comment]
I did a fresh checkout of `jgfouca/rrtmgp_interface_back_to_kokkos` branch.

**Review by bartgol — APPROVED** [review]

**Pull request merged**
